Generic roundtrip test

curino opened the pull request:
Relates to: #5

jacques-n commented on core/src/test/java/io/substrait/type/proto/GenericRoundtripTest.java:
```diff
@@ -0,0 +1,129 @@
+        } else if (type.equals(BigDecimal.class)) {
+            return BigDecimal.valueOf(rand.nextLong());
+        } else if (type.equals(Instant.class)){
+            return Instant.now();
```
same as comment above

curino commented on core/src/test/java/io/substrait/type/proto/GenericRoundtripTest.java:
```diff
@@ -0,0 +1,129 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class GenericRoundtripTest {
```
100% this was just a quick draft for me to poke at things and demonstrate the idea.

curino replied:
As soon as I find some time I will address the comments and post a new version. Any thoughts on how to extend this past base types? I.e., do you have any easy to use infra to create example expression etc? (maybe from some parsed SQL or else)?

curino commented on core/src/test/java/io/substrait/type/proto/GenericRoundtripTest.java:
```diff
@@ -0,0 +1,129 @@
+
+
+    private Object valGenerator(Class<?> type) {
+        Random rand = new Random();
```
I will look into it.

curino commented:
@jacques-n I fixed a good chunk of what you asked. 

Specifically the parametrization, and using junit reporting for uncovered cases. 

I also found and fixed the bug we had on round-tripping nullable fields (we missed to propagate the nullability in `ExpressionProtoConverter` for all/most types, it is now correct for the fields we cover and others more complex expressions don't seem to have nullability). 

I still need to address the comments around randomization not being tracked hence making repeatability of failure a pain (if you have pointers to good example let me know, otherwise I will look around).

curino commented:
I think I found and fixed another bug in Expression.toBytes() where we were missing to reset the `ByteBuffer` position and we generated an empty buffer as a copy (making the roundtrip for UUIDs not working). 

@jacques-n  let me know if you are ok to accept the 2 bugfixes and test collateral changes in one PR or prefer me to split them.

jacques-n commented on core/src/main/java/io/substrait/expression/Expression.java:
```diff
@@ -296,6 +296,7 @@ public ByteString toBytes() {
       ByteBuffer bb = ByteBuffer.allocate(16);
       bb.putLong(value().getMostSignificantBits());
       bb.putLong(value().getLeastSignificantBits());
+      bb.rewind();
```
Should this be `bb.flip()`?

curino replied:
Good point, I think in this case doesn't matter as we have a 16bytes and put in 2 longs, but generally I think `flip()` is likely the right answer. Updated.

curino replied:
BTW I still need to look at how to make the rand() more trackable. I found something in junit to randomize the order of method execution, but nothing super nice/obvious to me to track this rand generation. @jacques-n  do you have a suggestion for that?

jacques-n replied:
For now, let's just use a defined seed for all the random inputs so we avoid randomness. We can try to find something else later.

jacques-n commented on core/src/test/java/io/substrait/type/proto/GenericRoundtripTest.java:
```diff
@@ -0,0 +1,145 @@
+public class GenericRoundtripTest {
+    static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(GenericRoundtripTest.class);
+
+    static Random rand = new Random(123);
```
Let's do `private final` instead of static. Random is not thread safe so best to avoid accidentally sharing.

jacques-n replied:
Also best to add a comment that the value is not meaningful, just done to make sure test is repeatable.

curino replied:
I guess you already merged. Should I send in an upate? (Or we can do it in future iterations, as I assume we should expand this to also cover complex types).

jacques-n reviewed (approved):
Thanks for the contribution @curino !

curino commented:
Thanks @jacques-n  for the prompt and useful feedback! Looking forward to slowly contribute more and more to the project!
